File: day54/singleNumber.java
class Solution {
    public int singleNumber(int[] nums) {
        int single = 0;
        for (int num : nums)
            single ^= num;
        
        return single;
    }
}

// class Solution {
//     public int singleNumber(int[] nums) {
//         Map<Integer, Integer> map = new HashMap<>();

//         for (int num : nums) {
//             map.put(num, 1 + map.getOrDefault(num, 0));
//         }
        
//         for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
//             if (entry.getValue() == 1) return entry.getKey();
//         }

//         throw new NumberFormatException();
//     }
// }

// class Solution {
//     public int singleNumber(int[] nums) {
//         Arrays.sort(nums);
//         int i = 0;
//         while (i < nums.length - 1) {
//             if (nums[i] == nums[i + 1]) {
//                 i += 2;
//             } else {
//                 return nums[i];
//             }
//         }
//         return nums[nums.length - 1];
//     }
// }

File: day54/singleNumber.py
'''
136. 只出现一次的数字
给定一个非空整数数组，除了某个元素只出现一次以外，其余每个元素均出现两次。找出那个只出现了一次的元素。
说明：你的算法应该具有线性时间复杂度。 你可以不使用额外空间来实现吗？

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/single-number
'''

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        nums.sort()
        for i in range(0, len(nums) - 1, 2):
            if nums[i] != nums[i + 1]: return nums[i]
        return nums[-1]

# class Solution:
#     def singleNumber(self, nums: List[int]) -> int:
#         single = 0
#         for num in nums:
#             single ^= num
#         return single
